File: precog/php/src/Precog.php
<?php
/**
 * Provides access to the Precog API platform.
 *
 * Author: Alissa Pajer
 **/

define ("BASE_URL", "https://beta.precog.com");
define ("DEFAULT_VERSION", 1);

class PrecogAPI {

    public $apiKey = null;
    public $baseUrl = null;
    public $version = null;
    public $isError = false;
    public $errorMessage = null;

    /*
     * Initialize a new PrecogAPI object
     *
     * @param String $apiKey
     * @param String $baseUrl
     *
     */
    public function __construct($apiKey, $baseUrl = BASE_URL, $version = DEFAULT_VERSION)
    {
        $this->apiKey  = $apiKey;
        $this->baseUrl = self::cleanPath($baseUrl);
        $this->version = $version;
    }

    // ***************************
    // ****** ACCOUNTS APIS ******
    // ***************************
    public static function createAccount($email, $password, $baseUrl = BASE_URL, $version = DEFAULT_VERSION)
    {
        $url = self::baseActionUrl($baseUrl, $version, "accounts", "accounts");
        return self::baseRestHelper($url, json_encode(array("email"=>$email, "password"=>$password)), "POST");
    }

    public static function describeAccount($email, $password, $accountId, $baseUrl = BASE_URL, $version = DEFAULT_VERSION)
    {
        $url = self::baseActionUrl($baseUrl, $version, "accounts", "accounts").$accountId;
        return self::baseRestHelper($url, null, "GET", self::authHeaders($email, $password));
    }

       public static function addGrantToAccount($email, $password, $accountId, $grants, $baseUrl = BASE_URL, $version = DEFAULT_VERSION)
    {
        $url = self::baseActionUrl($baseUrl, $version, "accounts", "accounts/$accountId")."grants";
        return self::baseRestHelper($url, json_encode(array("grantId"=>$grants)), "POST", self::authHeaders($email, $password));
    }

    public static function describePlan($email, $password, $accountId, $baseUrl = BASE_URL, $version = DEFAULT_VERSION)
    {
        $url = self::baseActionUrl($baseUrl, $version, "accounts", "accounts/$accountId")."plan";
        return self::baseRestHelper($url, null, "GET", self::authHeaders($email, $password));
    }

    public static function changePassword($email, $oldPassword, $newPassword, $accountId, $baseUrl = BASE_URL, $version = DEFAULT_VERSION)
    {
        $url = self::baseActionUrl($baseUrl, $version, "accounts", "accounts/$accountId")."password";
        return self::baseRestHelper($url, json_encode(array("password"=>$newPassword)), "PUT", self::authHeaders($email, $oldPassword));
    }

    public static function deletePlan($email, $password, $accountId, $baseUrl = BASE_URL, $version = DEFAULT_VERSION)
    {
        $url = self::baseActionUrl($baseUrl, $version, "accounts", "accounts/$accountId")."plan";
        return self::baseRestHelper($url, null, "DELETE", self::authHeaders($email, $password));
    }

    public static function changePlan($email, $password, $accountId, $plan, $baseUrl = BASE_URL, $version = DEFAULT_VERSION)
    {
        $url = self::baseActionUrl($baseUrl, $version, "accounts", "accounts/$accountId")."plan";
        return self::baseRestHelper($url, json_encode(array("type"=>$plan)), "PUT", self::authHeaders($email, $password));
    }

    // ***************************
    // ****** INGEST APIS ********
    // ***************************

    public function ingest($path, $content, $type, $options = array() )
    {
        $contentType = strtolower($type);
        $parameters = array();

        switch($contentType) {
            case 'application/json':
            case 'json':
                $contentType = 'application/json';
            break;
            case 'text/csv':
            case 'csv':
                $contentType = 'text/csv';
                if(isset($options["delimiter"]))
                    $parameters["delimiter"] = $options["delimiter"];
                if(isset($options["quote"]))
                    $parameters["quote"] = $options["quote"];
                if(isset($options["escape"]))
                    $parameters["escape"] = $options["escape"];
            break;
            default:
                throw Error("argument 'type' must be 'json' or 'csv'");
        }

        if(isset($options["ownerAccountId"]))
            $parameters["ownerAccountId"] = $options["ownerAccountId"];

        $parameters["apiKey"] = $this->apiKey;
        $qsparams = array();
        foreach ($parameters as $parameter => $value) {
            $qsparams[] = $parameter."=".urlencode($value);
        }

        $url = $this->actionUrl("ingest", (isset($options["async"]) && $options["async"] ? "a" : "")."sync/fs"). self::cleanPath($path) ."?".implode("&", $qsparams);
        $return = $this->restHelper($url, $content, "POST", array("Content-Type" => $contentType));
        return $return;

    }

    /*
    * Record a new event
    *
    * @param String $path The path in which to store this event
    * @param Array $events event data
    *
    * @return Bool - success/failure
    */
    public function store($path, $event, $options = array())
    {
        return $this->ingest($path, json_encode($event), "application/json", $options);
    }

    public function delete($path)
    {
        $path2  = $this->actionUrl("ingest", "sync/fs") . self::cleanPath($path) . "?apiKey=" . $this->apiKey;
        $return = $this->restHelper($path2, null, "DELETE");
        return $return !== false;
    }

    // ***************************
    // ****** METADATA APIS ******
    // ***************************
    public function retrieveMetadata($path, $type = "")
    {
        $url = $this->actionUrl("meta", "fs") . self::cleanPath($path) . "?apiKey=".$this->apiKey."#".$type;
        $return = $this->restHelper($url, null, "GET");
        return $return;
    }

    /*
     * Returns an array of sub-paths
     * @params String - path
     *
     * @return Array - an array of values
     */
    public function listChildren($path)
    {
        $path = self::cleanPath($path);
        $path2  = $this->actionUrl("meta","fs")."$path?apiKey=" . $this->apiKey."#children";
        $return = $this->restHelper($path2, null, "GET");
        return $return ? $return['children'] : $return;
    }

    // ***************************
    // ****** ANALYTICS APIS *****
    // ***************************

    /*
     * Returns the value of the query
     * @params String - raw Quirrel
     *
     * @return Array - an array of values
     */

    public function query($quirrel, $options = array())
    {
        $params = array(
            "apiKey=" . $this->apiKey,
            "q=" . urlencode($quirrel)
        );
        if(isset($options["limit"])) {
            $params[] = "limit=" . $options["limit"];
        }

        if(isset($options["skip"])) {
            $params[] = "skip=" . $options["skip"];
        }
        if(isset($options["sortOn"])) {
            $params[] = "sortOn=" . urlencode(json_encode($options["sortOn"]));
            if(isset($options["sortOrder"])) {
                $params[] = "sortOrder=" . $options["sortOrder"];
            }
        }

        $path2  = $this->actionUrl("analytics", "fs")."?" .implode("&", $params);
        $return = $this->restHelper($path2, null, "GET");
        return $return;
    }

    // ***************************
    // ****** SECURITY APIS *****
    // ***************************
    public function listKeys()
    {
        $url = $this->actionUrl("security","apikeys" )."?apiKey=".$this->apiKey;
        $return = $this->restHelper($url, null, "GET");
        return $return;
    }

    public function createKey($grants)
    {
        $url = $this->actionUrl("security","apikeys")."?apiKey=".$this->apiKey;
        $return = $this->restHelper($url, json_encode($grants), "POST");
        return $return;
    }

    public function describeKey($apiKey)
    {
        $url = $this->actionUrl("security","apikeys").$apiKey."?apiKey=".$this->apiKey;
        $return = $this->restHelper($url, null, "GET");
        return $return;
    }

    public function deleteKey($apiKey)
    {
        $url = $this->actionUrl("security","apikeys").$apiKey."?apiKey=".$this->apiKey;
        $return = $this->restHelper($url, null, "DELETE");
        return $return !== false;
    }

    public function retrieveGrants($apiKey)
    {
        $url = $this->actionUrl("security","apikeys").$apiKey."/grants/?apiKey=".$this->apiKey;
        $return = $this->restHelper($url, null, "GET");
        return $return;
    }

    public function addGrantToKey($apiKey, $grant)
    {
        $url = $this->actionUrl("security","apikeys").$apiKey."/grants/?apiKey=".$this->apiKey;
        $return = $this->restHelper($url, json_encode(array("grantId"=>$grant)), "POST");
        return $return !== false;
    }

    public function removeGrant($apiKey, $grantId)
    {
        $url = $this->actionUrl("security","apikeys").$apiKey."/grants/".$grantId."?apiKey=".$this->apiKey;
        $return = $this->restHelper($url, null, "DELETE");
        return $return !== false;
    }

    public function createGrant($type)
    {
        $url = $this->actionUrl("security","grants")."?apiKey=".$this->apiKey;
        $return = $this->restHelper($url, json_encode($type), "POST");
        return $return;
    }

    public function describeGrant($grantId)
    {
        $url = $this->actionUrl("security", "grants").$grantId."?apiKey=".$this->apiKey;
        $return = $this->restHelper($url, null, "GET");
        return $return;
    }

    public function deleteGrant($grantId)
    {
        $url = $this->actionUrl("security","grants").$grantId."?apiKey=".$this->apiKey;
        $return = $this->restHelper($url, null, "DELETE");
        return $return !== false;
    }

    public function listGrantChildren($grantId)
    {
        $url = $this->actionUrl("security","grants").$grantId."/children/?apiKey=".$this->apiKey;
        $return = $this->restHelper($url, null, "GET");
        return $return;
    }

    public function createGrantChild($grantId, $type)
    {
        $url = $this->actionUrl("security","grants").$grantId."/children/?apiKey=".$this->apiKey;
        $return = $this->restHelper($url, json_encode($type), "POST");
        return $return;
    }

    /*********************************
     **** PRIVATE helper function ****
     *********************************/
    private function restHelper($resturl, $params = null, $verb = 'GET', $headers = false) {
        $result = self::baseRestHelper($resturl, $params, $verb, $headers);
        if($result['ok']) {
            $this->isError = false;
            $this->errorMessage = null;
            return $result['data'];
        } else {
            $this->isError = true;
            $this->errorMessage = isset($result['error']) ? $result['error'] : "an error occurred";
            return false;
        }
    }

    private static function baseRestHelper($resturl, $params = null, $verb = 'GET', $headers = false) {
echo("$verb $resturl\n");
//if($params) var_dump($params);
        $return = array('ok' => true);
        $http_params = array(
            'http' => array(
                'method'        => $verb,
                'ignore_errors' => false,
                'header'        => $headerString = self::getHeaderString($headers)
        ));
//var_dump($headerString);

        // workaround for php bug where http headers don't get sent in php 5.2
        if(version_compare(PHP_VERSION, '5.2.14') < 0) {
            ini_set('user_agent', 'PHP-SOAP/' . PHP_VERSION . "\r\n" . $headerString);
        }

        if ($params !== null && ($verb == 'POST' || $verb == 'PUT')) {
            $http_params['http']['content'] = $params;
        }

        $stream_context = stream_context_create($http_params);
        $file_pointer = @fopen($resturl, 'rb', false, $stream_context);
        if (!$file_pointer) {
            $stream_contents = false;
        } else {
            $stream_meta_data = stream_get_meta_data($file_pointer);
            $stream_contents = stream_get_contents($file_pointer);
            fclose($file_pointer);
        }
//if(isset($stream_contents)) var_dump($stream_contents);
//if(isset($stream_meta_data)) var_dump($stream_meta_data);


        if ($verb==="DELETE" || $stream_contents !== false) {

            /*
             * In the case of we're receiving stream data back from the API,
             * json decode it here.
             */
            if (strlen($stream_contents) > 0) {
                $result = json_decode($stream_contents, true);

                if ($result === null) {
                    error_log("Exception:  " . $stream_contents);
                } else {
                    $return['data'] = $result;
                }
            /*
             * In the case of posting data (recordEvent) the API will return a 0
             * length response, in this scenario we're looking for the http 200
             * header code to indicate the data was successfully received.
             */
            } else {
                if (self::checkOkResponse($stream_meta_data['wrapper_data'][0])) {
                    $return['data'] = true;
                } else {
                    $return['ok'] = false;
                }
            }

        } else {
            /*
             * If there's an error message in the response
             * headers...send that back to the user
             */
            if (isset($http_response_header[0])) {
                $return['ok'] = false;
                $return['error'] = $http_response_header[0];
            } else {
                throw new Exception("$verb $resturl failed");
            }

        }
        return $return;
    }

    static private function checkOkResponse($response) {
        return
               stripos($response, "200") !== false
            || stripos($response, "201") !== false
            || stripos($response, "202") !== false
            || stripos($response, "204") !== false
        ;
    }

    private function actionUrl($service, $action = false) {
       return self::baseActionUrl($this->baseUrl, $this->version, $service, $action);
    }

    private static function getHeaderString($headers) {
        if(!$headers) {
            $headers = array();
        }
        if(!isset($headers['Content-Type']))
            $headers['Content-Type'] = "application/json";
        $result = array();
        foreach ($headers as $key => $value) {
            $result[] = "$key: $value";
        }
        return implode("\r\n", $result);
    }

    private static function baseAuth($user, $password) {
        $tok = "$user:$password";
        return "Basic ".base64_encode($tok);
    }

    private static function cleanPath($path) {
        return trim($path, '/');
    }

    private static function baseActionUrl($baseUrl, $version, $service, $action = false) {
        return $baseUrl."/".$service."/v".$version."/".($action ? $action."/" : "");
    }

    private static function authHeaders($user, $password) {
        return array("Authorization" => self::baseAuth($user, $password));
    }
}
?>

File: precog/ruby/lib/precog.rb
# Copyright (C) 2012 by Precog, Inc. All rights reserved.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# No portion of this Software shall be used in any application which does not
# use the Precog platform to provide some subset of its functionality.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

require 'logger'
require 'net/http'
require 'net/https'

require 'pp'
require 'time'
require 'uri'

require 'rubygems'
require 'json'

module Precog

  # API server constants
  module API
    HOST = 'devapi.precog.com'
    PORT = 443
  end

  module Paths
    FS = '/fs'
  end

  # Services constants
  module Services
    ANALYTICS = '/analytics/v1'
    ACCOUNTS = '/accounts/v1'
    INGEST = '/ingest/v1'
  end

  
  
  # Base exception for all Precog errors
  class PrecogError < StandardError
    def intialize(message)
      super
    end
  end

  # Raised on HTTP response errors
  class HttpResponseError < PrecogError
    attr_reader :code
    def initialize(message, code)
      super(message)
      @code = code
    end
  end

  # Simple HTTP client for Precog API
  class HttpClient

    # Initialize an HTTP connection
    def initialize(api_key, host, port)
      @api_key    = api_key
      @host       = host
      @port       = port
      @conn       = Net::HTTP.new(host, port)
      if @port == 443
        @conn.use_ssl = true
        @conn.verify_mode = OpenSSL::SSL::VERIFY_NONE
      end
    end

    def basic_auth(user, password)
      { :auth =>{ :user=>user, :password=>password }}
    end

    def action_url(service, action )
       "#{service}/#{action}"
    end

    # Sanitize a URL path
    def sanitize_path(path)
      newpath = path.gsub(/\/+/, '/')
      while newpath.gsub!(%r{([^/]+)/\.\./?}) { |match|
        $1 == '..' ? match : ''
      } 
      end
      newpath.gsub(%r{/\./}, '/').sub(%r{/\.\z}, '/')
    end

    # Properties must always be prefixed with a period
    def sanitize_property(property)
      if property && !property.start_with?('.')
        property = ".#{property}"
      end
      property
    end
    

    # Send an HTTP request
    def method_missing(name, service, action, options={}, content_type = 'application/json')
      options[:body]    ||= ''
      options[:headers] ||= {}
      options[:parameters] ||= {}
      options[:auth] ||= {}

      # Add api key to path and set headers
      path = action_url(service,action)
      path = sanitize_path(path)

     
      if !@api_key.nil? && !@api_key.empty?
        path +="?apiKey=#{@api_key}"
      end

      options[:parameters].each do |k, v|
        path += "&#{k}=#{URI.escape(v)}"
      end

      options[:headers].merge!({'Content-Type' => content_type})

      # Set up message
      message = "#{name.to_s.upcase} to #{@host}:#{@port}#{path} with headers (#{options[:headers].to_json })"
      if options[:body]
        body = (content_type=='application/json')? options[:body].to_json : options[:body]
        message += " and body (#{body})"
      end

      # Send request and get response
      begin
        request = Net::HTTP.const_get(name.to_s.capitalize).new(path)
        if options[:auth]
          request.basic_auth options[:auth][:user],options[:auth][:password]
        end
        options[:headers].each { |k,v| request.add_field(k, v) }
        response = @conn.request(request, body)
      rescue StandardError => e
        message += " failed (#{e.inspect})"
        raise HttpResponseError.new(message, 500)
      end

      # Check HTTP status code
      if ![200,202].include?(response.code.to_i)
        message += " returned non-200 status (#{response.code}): #{response.inspect}"
        raise HttpResponseError.new(message, response.code.to_i)
      end

      # Try parsing JSON response
      begin
        response_data = JSON.parse(response.body)
      rescue StandardError => e
        response_data = response.body
        if response_data.start_with?('"') && response_data.end_with?('"')
          response_data = response_data[1...-1]
        end
      end

      response_data
    end
  end

  # Precog base class
  class Precog

    # Initialize an API client
    def initialize(api_key, host = API::HOST, port = API::PORT)
      @api_key = api_key
      @api = HttpClient.new(api_key, host, port)
    end

    def self.from_heroku(token)
      values=Utils.from_token(token)
      no_key_api=Precog.new(nil, values[:host])
      response =no_key_api.describe_account(values[:user],values[:pwd],values[:account_id])
      Precog.new(response['apiKey'], values[:host])
    end

    ######################
    ###    ACCOUNTS    ###
    ######################

    #Creates a new account ID, accessible by the specified email address and password, or returns the existing account ID.
    def create_account(email, password)
      @api.post(Services::ACCOUNTS,"accounts/",:body => { :email=> email, :password=> password } )
    end
    
    #Retrieves the details about a particular account. This call is the primary mechanism by which you can retrieve your master API key.
    def describe_account(email, password, account_id)
      @api.get(Services::ACCOUNTS,"accounts/#{account_id}",@api.basic_auth(email, password) )
    end

    #Adds a grant to an account's API key.
    def add_grant_to_account(email, password, account_id, grant_id)
      @api.post(Services::ACCOUNTS,"accounts/grants/",@api.basic_auth(email, password).merge!({ :body => { :grantId => grant_id  } }))
    end

    #Describe Plan
    def describe_plan(email, password, account_id)
      @api.get(Services::ACCOUNTS,"accounts/#{account_id}/plan",@api.basic_auth(email, password))
    end

    #Changes an account's plan (only the plan type itself may be changed). Billing for the new plan, if appropriate, will be prorated.
    def change_plan(email, password, account_id, type)
      @api.put(Services::ACCOUNTS,"accounts/#{account_id}/plan",@api.basic_auth(email, password).merge!({ :body => { :type => type } }))
    end

    #Changes your account access password. This call requires HTTP Basic authentication using the current password.
    def change_password(email, password, account_id, new_password)
      @api.put(Services::ACCOUNTS,"accounts/#{account_id}/password", @api.basic_auth(email, password).merge!({:body => { :password => new_password  } }))
    end

    #Deletes an account's plan. This is the same as switching a plan to the free plan.
    def delete_plan(email, password, account_id)
      @api.delete(Services::ACCOUNTS,"accounts/#{account_id}/plan",@api.basic_auth(email, password))
    end

    ######################
    ###     INGEST     ###
    ######################
    def get_api_key(options)
      { :apiKey => (options['apiKey']) || @api_key }
    end


    # Ingests csv or json data at the specified path
    # (mode = batch, receipt = true), 
    # (mode = batch, receipt = false), 
    # (mode = streaming)
    def ingest(path, content, type, options={})
      if !content
        raise Error.new("argument 'content' must contain a non empty value formatted as described by type")
      end

      parameters={}
      case(type.downcase) 
        when 'application/x-gzip','gz','gzip' then
          type = 'application/x-gzip'
        when 'zip' then 
          type = 'application/zip'
        when 'application/json','json' then
          type = 'application/json'
        when 'text/csv','csv' then
          type = 'text/csv'
          if options[:delimiter]
            parameters['delimiter'] = options[:delimiter]
          end
          if options[:quote]
            parameters['quote'] = options[:quote]
          end
          if options[:escape]
            parameters['escape'] = options[:escape]
          end
        else
          raise "argument 'type' must be 'json' or 'csv'"
      end
      
      parameters['mode']=options[:mode]
      if options[:mode]=="batch"
        parameters['receipt']="#{options[:receipt]}"
      end

      if options[:owner_account_id]
          parameters['owner_account_id'] = options[:owner_account_id]
      end
      action = @api.sanitize_path("/#{Paths::FS}/#{path}")
      @api.post(Services::INGEST,action, 
        { :parameters=> parameters, :body => content },type)
    end

    def ingest_batch(path,content,type, receipt, options={})
      ingest(path, content, type, options.merge!({:mode=> 'batch' , :receipt=> receipt}))
    end

    def ingest_stream(path,content,type, options={})
      ingest(path, content, type, options.merge!({:mode=> 'streaming' }))
    end

    # Store a record at the specified path
    def store(path, event, options = {})
        ingest(path, event, "json", options.merge!({:mode=> 'batch' , :receipt=> 'true'}))
    end


    def delete(path)
      action = @api.sanitize_path("#{Paths::FS}/#{path}")
      @api.delete(Services::INGEST,action)
    end

    # Send a quirrel query to be evaluated relative to the specified base path.
    def query(path, query)
      path = "#{Paths::FS}/#{path}" unless path.start_with?(Paths::FS)
      path = @api.sanitize_path(path)
      options={  :parameters => { :q => query } } 
      @api.get(Services::ANALYTICS, path, options)
    end

    private :ingest

  end

  class Utils
    def self.to_token(user,pwd,host, account_id, api_key, root_path)
      Base64.encode64("#{user}:#{pwd}:#{host}:#{account_id}:#{api_key}:#{root_path}").gsub(/\n/, '')
    end

    def self.from_token(token)
      values=Base64.decode64(token).split(":")
      { :user=>values[0], :pwd=>values[1], :host=>values[2], :account_id=>values[3], :api_key=> values[4], :root_path=> values[5] }
    end
  end
end
